fix: WASM provided by nix build of barretenberg

kobyhallx opened the pull request:
# Related issue(s)
    
#23     
# Resolves # <!-- link to issue -->

# Description

Switching Noir to barretenberg upstream introduced requirement for Barretenberg WASM to be located at `$BARRETENBERG_BIN_DIR`. Absence of this variable breaks builds.
    
## Summary of changes

libbarretenberg-wasm32 is built by nix from upstream AztecProtocol/barretenberg, uploaded to artifacts and later downloaded and reused by workflows.    
    
## Dependency additions / changes

libbarretenberg-wasm32 
    
## Test additions / changes
N/A    

# Checklist
    
- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.
    
# Additional context

Commit:
fix: Use wasm from nix build

Commit:
fix: adds missing nix installation

Commit:
dbg: should contain result from nix

Commit:
dbg: add backslash to upload dir

Commit:
dbg: readlink workaround

Commit:
dbg: download path

Commit:
dbg: clearer download path

Commit:
dbg: setting bb path to download location

Commit:
dbg: test workflow-run

Commit:
dbg: correct workflow name

Commit:
dbg: change name

Commit:
dbg: typos fix

Commit:
enable custom branch

Commit:
dbg: all build bb-wasm

Commit:
dbg: needs added

Commit:
fix(build): bb wasm based on nix

phated commented on .github/workflows/publish-apple-darwin-wasm.yml:
```diff
@@ -11,8 +20,36 @@ on:
+      # https://github.com/actions/upload-artifact/issues/92#issuecomment-1080347032    
+      - name: Build barretenberg as libbarretenberg-wasm32
+        run: |
+          nix build "git+https://github.com/AztecProtocol/barretenberg#wasm32"
```
This won't be pinned to the bb commit that noir is pinned to. We need to be using the lockfile that noir is

kobyhallx replied:
which lockfile?

phated replied:
https://github.com/noir-lang/noir/blob/master/flake.lock#L15-L17

kobyhallx replied:
pinning done in nix-barretenberg.yml

phated commented on .github/workflows/publish-apple-darwin-wasm.yml:
```diff
@@ -11,8 +20,36 @@ on:
+          nix build "git+https://github.com/AztecProtocol/barretenberg#wasm32"
+          echo "ARTIFACT_UPLOAD_PATH=$(readlink -f result)" >> $GITHUB_ENV
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: libbarretenberg-wasm32
+          path: ${{ env.ARTIFACT_UPLOAD_PATH }}
+          retention-days: 3
```
What's the reason for making this a separate job and then uploading the artifact? Could we combine into one job and not upload?

kobyhallx replied:
in separate update to be converted to workflow_run invocation to share one build across all platform workflows

phated replied:
Isn't the wasm pre-built in the cachix binary cache though? I'd expect us to just be downloading the wasm file

kobyhallx replied:
it's not only that, windows won't run nix and needs to download artifcat from somewhere anyway

phated reviewed (changes requested):
This is a great solution. We need to figure out how to pin bb to the commit noir uses though, perhaps using the lockfile?

Commit:
fix: barretenberg pinning

- Barretenberg pinned in nix-barretenberg.yml
- Barretenberg artfcats uploaded to be shared across workflows
- system specific workflows triggered by workflow_run - sucess

TomAFrench commented on .github/workflows/nix-barretenberg.yml:
```diff
@@ -0,0 +1,52 @@
+    - name: Collect locked barretenberg rev
+      run: |
+        echo "BB_REV=$(jq -r .nodes.barretenberg.locked.rev ./noir/flake.lock)" >> $GITHUB_ENV
+  
```
```suggestion
```

nit

phated commented on .github/workflows/publish-apple-darwin-wasm.yml:
```diff
@@ -1,14 +1,12 @@
-    - cron: "0 2 * * *" # run at 2 AM UTC
-  push:
+on: 
+  workflow_run:
```
This doesn't trigger a workflow when another runs, it is used to make re-usable workflows that can be called as jobs.

kobyhallx replied:
as specified line below this will get triggered by `libbarretenberg` workflow completion

phated replied:
~~No, that's not how these work~~

I was incorrect, these only work on default branch, which @kobyhallx forked and test. See action runs at https://github.com/kobyhallx/build-nargo-nl/actions

TomAFrench committed:
Update .github/workflows/nix-barretenberg.yml

phated commented on .github/workflows/publish-apple-darwin-wasm.yml:
```diff
@@ -45,13 +43,21 @@ jobs:
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
+        uses: dawidd6/action-download-artifact@v2
```
Can we use a github official action or not? https://github.com/actions/download-artifact

kobyhallx replied:
we tried, but official one does not work cross workflows as seen on 
https://github.com/actions/download-artifact/issues/172

phated reviewed (changes requested):
Question about your choice about an unofficial action when an official one exists. Maybe there is a reason we can't use official version?

phated commented on .github/workflows/publish-apple-darwin-wasm.yml:
```diff
@@ -45,13 +43,21 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
```
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

phated commented on .github/workflows/publish-linux.yml:
```diff
@@ -48,13 +45,22 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download pre-built libbarretenberg-wasm32
```
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

phated commented on .github/workflows/publish-x86_64-pc-windows-wasm.yml:
```diff
@@ -40,13 +36,21 @@ jobs:
             target/
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
+      - name: Download artifact
```
```suggestion
      # We use an unofficial download plugin to support cross-workflow downloads
      # Ref https://github.com/actions/download-artifact/issues/172
      - name: Download artifact
```

phated reviewed (changes requested):
Let's add this comment to each workflow.

phated committed:
Apply suggestions from code review

phated committed:
format workflows

phated reviewed (approved):
We need to update Noir to dispatch to this new workflow.
